### PR TITLE
pdf sparse honors anal.jmpmid + afb+ fills op_pos array (x86 only) ##anal

### PR DESCRIPTION
**Pull request by kazarmy**

### CODE REVIEW

File: libr/anal/fcn.c [hunk 1]
```diff
@@ -907,7 +908,8 @@ static int fcn_recurse(RAnal *anal, RAnalFunction *fcn, ut64 addr, ut8 *buf, ut6
 		return R_ANAL_RET_ERROR; // MUST BE TOO DEEP
 	}
 
-	if (r_anal_get_fcn_at (anal, addr, 0)) {
+	fcn_at_addr = r_anal_get_fcn_at (anal, addr, 0);
```
Comment (radare) [inline libr/anal/fcn.c]: Define and assign in the same line, same for bbuf

File: libr/anal/fcn.c [hunk 2]
```diff
@@ -1797,9 +1799,11 @@ R_API RAnalFunction *r_anal_fcn_find_name(RAnal *anal, const char *name) {
 /* rename RAnalFunctionBB.add() */
 R_API int r_anal_fcn_add_bb(RAnal *anal, RAnalFunction *fcn, ut64 addr, ut64 size, ut64 jump, ut64 fail, int type, RAnalDiff *diff) {
 	RAnalBlock *bb = NULL, *bbi;
+	ut8 *bbuf;
 	RListIter *iter;
 	bool mid = false;
 	st64 n;
+	int old_num_bbs;
```
Comment (radare) [inline libr/anal/fcn.c]: And here

File: libr/anal/fcn.c [hunk 3]
```diff
@@ -1818,14 +1822,22 @@ R_API int r_anal_fcn_add_bb(RAnal *anal, RAnalFunction *fcn, ut64 addr, ut64 siz
...
+		fcn_recurse (anal, fcn, addr, bbuf, MAXBBSIZE, 1);
+		r_anal_fcn_update_tinyrange_bbs (fcn);
+		free (bbuf);
+	}
+	if (old_num_bbs >= r_list_length (fcn->bbs)) {
+		eprintf ("fcn_recurse failed\n");
+		return false;
 	}
-	bb->addr = addr;
+	bb = r_list_tail (fcn->bbs)->data;
```
Comment (radare) [inline libr/anal/fcn.c]: List can be empty?

File: libr/anal/fcn.c (hunk 1 above)
Comment (kazarmy) [reply libr/anal/fcn.c]: ok 6dcfed1ae1f16c9d345bdc36975b341c1b6bb04d

File: libr/anal/fcn.c (hunk 2 above)
Comment (kazarmy) [reply libr/anal/fcn.c]: no longer relevant due to dcd1f66ae9ee3cdadee69419ac4a2b29098e6c0b

File: libr/anal/fcn.c (hunk 3 above)
Comment (kazarmy) [reply libr/anal/fcn.c]: no longer relevant due to dcd1f66ae9ee3cdadee69419ac4a2b29098e6c0b

File: libr/anal/fcn.c [hunk 4]
```diff
@@ -946,7 +947,7 @@ static int fcn_recurse(RAnal *anal, RAnalFunction *fcn, ut64 addr, ut8 *buf, ut6
 		if (r_cons_is_breaked ()) {
 			break;
 		}
-		if ((len - addrbytes * idx) < 5) {
+		if ((len - addrbytes * idx) < 5 && len == MAXBBSIZE) { // Should use opt.bb_max_size here
```
Comment (radare) [inline libr/anal/fcn.c]: the maxbbsize thing should die

Comment (kazarmy) [reply libr/anal/fcn.c]: added TODO bbd885b14e5bd1eb998c2c78068a051f6edb2880

Comment (radare) [reply libr/anal/fcn.c]: https://github.com/radare/radare2/issues/9564

Comment (kazarmy) [reply libr/anal/fcn.c]: i'll have a look

**Comment by radare** [comment]
Just a reminder for the future. No need to kill maxbbsz here

File: libr/anal/fcn.c [hunk 5]
```diff
@@ -1818,14 +1820,38 @@ R_API int r_anal_fcn_add_bb(RAnal *anal, RAnalFunction *fcn, ut64 addr, ut64 siz
...
 		}
 	}
-	if (!bb) {
-		bb = appendBasicBlock (anal, fcn, addr);
+	if (x86) {
+		if (bb) {
+			r_list_delete_data (fcn->bbs, bb);
+		}
+		ut8 *bbuf = malloc (size);
+		if (bbuf) {
```
Comment (ret2libc) [inline libr/anal/fcn.c]: we should probably just return from the function if we can't even malloc something

File: libr/anal/fcn.c [hunk 6]
```diff
@@ -1818,14 +1820,38 @@ R_API int r_anal_fcn_add_bb(RAnal *anal, RAnalFunction *fcn, ut64 addr, ut64 siz
...
+			anal->iob.read_at (anal->iob.io, addr, bbuf, size);
+			fcn_recurse (anal, fcn, addr, bbuf, size, 1);
+			r_anal_fcn_update_tinyrange_bbs (fcn);
+			free (bbuf);
+		}
+		bb = NULL;
+		r_list_foreach (fcn->bbs, iter, bbi) {
```
Comment (ret2libc) [inline libr/anal/fcn.c]: use `r_anal_fcn_bbget_at`. it should be what you are looking for.

Comment (kazarmy) [reply libr/anal/fcn.c]: ok 4a25925a379ffc01c94504dd733df6dddd532348

File: libr/anal/fcn.c (hunk 5 above)
Comment (kazarmy) [reply libr/anal/fcn.c]: ok 54813aae942774a0a606310a6823ce19c95b036b

**Comment by kazarmy** [comment]
next pr will probably kill the
https://github.com/radare/radare2/blob/bbd885b14e5bd1eb998c2c78068a051f6edb2880/libr/anal/fcn.c#L953

File: libr/anal/fcn.c [hunk 7]
```diff
@@ -1818,14 +1820,34 @@ R_API int r_anal_fcn_add_bb(RAnal *anal, RAnalFunction *fcn, ut64 addr, ut64 siz
...
+	if (x86) {
+		if (bb) {
+			r_list_delete_data (fcn->bbs, bb);
+		}
+		ut8 *bbuf = malloc (size);
+		if (!bbuf) {
+			eprintf ("malloc failed\n");
+			return false;
+		}
+		anal->iob.read_at (anal->iob.io, addr, bbuf, size);
```
Comment (radare) [inline libr/anal/fcn.c]: check for return, or use calloc and cast (void) in this call

**Pull request merged**